tree/ntuple/v7/inc/ROOT/RFieldVisitor.hxx

```diff
@@ -0,0 +1,122 @@
+};
+   
+// clang-format off
+/**
+\class ROOT::Experimental::RPrepareVisitor
+\ingroup NTuple
+\brief Visitor used for a prepare run to collect information needed by another visitorclass.
+
+ Currently used for RPrintVisitor in RNTupleReader::Print() to collect information about levels, maximal depth etc.
+*/
+// clang-format on
+   
+class RPrepareVisitor: public RNTupleVisitor {
+private:
+   int fDeepestLevel;
+   int fNumFields;
+public:
+   RPrepareVisitor(int deepestLevel=0, int numFields=0): fDeepestLevel{deepestLevel}, fNumFields{numFields} { }
+   void visitField(const Detail::RFieldBase &/*field*/, int level) override;
+   void visitRootField(const RFieldRoot &/*field*/, int /*level*/) override { }
+   int getDeepestLevel() const {return fDeepestLevel;}
+   int getNumFields() const {return fNumFields;}
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::RPrintVisitor
+\ingroup NTuple
+\brief Contains settings for printing and prints a summary of an RField instance.
+     
+ Instances of this class are currently only invoked by RNTupleReader::Print() -> RFieldBase::AcceptVisitor()
+*/
+// clang-format on
+
+class RPrintVisitor : public RNTupleVisitor {
+private:
+   /// Holds std::ostream which is used for holding the printed output
+   std::ostream &fOutput;
+   char fFrameSymbol;
+   /// Indicates maximal number of allowed characters per line
+   int fWidth;
+   int fDeepestLevel;
+   int fNumFields;
+   int fAvailableSpaceKeyString;
+   int fAvailableSpaceValueString;
+   /// Keeps track when | is used for the tree-like structure.
+   std::vector<bool> fFlagforVerticalLines;
+   /// KeyString refers to the left side containing the word "Field" and its hierarchial order
+   std::string KeyString(const Detail::RFieldBase &field, int level);
+   /// ValueString refers to the right side containing the type and name
+   std::string ValueString(const Detail::RFieldBase &field);
+public:
+   RPrintVisitor(std::ostream &out = std::cout, char fillSymbol = '*', int width = 80, int deepestLevel = 1, int numFields = 1)
+   : fOutput{out}, fFrameSymbol{fillSymbol}, fWidth{width}, fDeepestLevel{deepestLevel}, fNumFields{numFields}
+   { }
+   /// Prints summary of Field
+   void visitField(const Detail::RFieldBase &field, int level) override;
+   void visitRootField(const RFieldRoot &/*field*/, int /*level*/) override { };
+   void SetFrameSymbol(char s) {fFrameSymbol = s;}
+   void SetWidth(int w) {fWidth = w;}
+   void SetDeepestLevel(int d) {fDeepestLevel = d;}
+   void SetNumFields(int n) {fNumFields = n;}
+   void SetAvailableSpaceForStrings() {
+      fAvailableSpaceKeyString = std::min(4*fDeepestLevel+4+static_cast<int>(std::to_string(fNumFields).size()), fWidth-15);
+      fAvailableSpaceValueString = fWidth - 6 - fAvailableSpaceKeyString;
+   }
+   void ResizeFlagVec() {fFlagforVerticalLines.resize(fDeepestLevel-1);}
+   void ResizeFlagVec(int size) {fFlagforVerticalLines.resize(size);}
```

Comment:
It's used for unit tests.

Reply:
I see, still we should find another way that does not make the methods part of the public interface. We could, for instance, add the unit test as friend test to the class so that the test can manipulate private data members.